File: src/web/routes/browse.py
"""Filesystem directory browser for the studio UI."""

from pathlib import Path

from fastapi import APIRouter, HTTPException, Query

from src.config import get_settings

router = APIRouter(prefix="/api/browse", tags=["browse"])

VIDEO_EXTS = {".mp4", ".mov", ".avi", ".mkv", ".webm", ".m4v", ".wmv"}


def _is_within_roots(target: Path, roots: list[Path]) -> bool:
    for root in roots:
        try:
            target.relative_to(root)
            return True
        except ValueError:
            continue
    return False


@router.get("")
async def browse_directory(
    path: str = Query(default="~", description="Absolute path to list"),
) -> dict:
    """List subdirectories and video files at the given path."""
    settings = get_settings()
    target = Path(path).expanduser().resolve()

    if not _is_within_roots(target, settings.browse_roots):
        raise HTTPException(
            status_code=403,
            detail="Path is outside configured browse roots",
        )

    if not target.exists():
        raise HTTPException(status_code=404, detail=f"Path not found: {target}")
    if not target.is_dir():
        raise HTTPException(status_code=400, detail=f"Not a directory: {target}")

    dirs: list[dict] = []
    files: list[dict] = []

    try:
        for entry in sorted(target.iterdir(), key=lambda e: e.name.lower()):
            # Skip hidden files/dirs.
            if entry.name.startswith("."):
                continue
            if entry.is_dir():
                dirs.append({"name": entry.name, "path": str(entry), "type": "dir"})
            elif entry.suffix.lower() in VIDEO_EXTS:
                files.append({"name": entry.name, "path": str(entry), "type": "file"})
    except PermissionError:
        raise HTTPException(status_code=403, detail=f"Permission denied: {target}")

    return {
        "current": str(target),
        "parent": str(target.parent) if target.parent != target else None,
        "dirs": dirs,
        "files": files,
        "video_count": len(files),
    }

File: apps/api/src/routes/browse.ts
import { existsSync, readdirSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { Hono } from "hono";
import { getSettings, HttpError } from "@ave/core";

const VIDEO_EXTS = new Set([
  ".mp4",
  ".mov",
  ".avi",
  ".mkv",
  ".webm",
  ".m4v",
  ".wmv",
]);

function isWithinRoots(target: string, roots: string[]): boolean {
  const normalizedTarget = resolve(target);
  return roots.some((root) => {
    const normalizedRoot = resolve(root);
    return (
      normalizedTarget === normalizedRoot ||
      normalizedTarget.startsWith(`${normalizedRoot}${process.platform === "win32" ? "\\" : "/"}`)
    );
  });
}

export const browseRoutes = new Hono();

browseRoutes.get("/", (c) => {
  const settings = getSettings();
  const rawPath = c.req.query("path") ?? "~";
  const expanded = rawPath.startsWith("~")
    ? rawPath.replace(/^~/, process.env.HOME || process.env.USERPROFILE || "")
    : rawPath;
  const target = resolve(expanded);

  if (!isWithinRoots(target, settings.browseRoots)) {
    throw new HttpError(403, "Path is outside configured browse roots");
  }
  if (!existsSync(target)) {
    throw new HttpError(404, `Path not found: ${target}`);
  }
  if (!statSync(target).isDirectory()) {
    throw new HttpError(400, `Not a directory: ${target}`);
  }

  const dirs: Array<{ name: string; path: string; type: "dir" }> = [];
  const files: Array<{ name: string; path: string; type: "file" }> = [];

  try {
    for (const entry of readdirSync(target).sort((a, b) =>
      a.localeCompare(b, undefined, { sensitivity: "base" }),
    )) {
      if (entry.startsWith(".")) continue;
      const fullPath = resolve(target, entry);
      const stats = statSync(fullPath);
      if (stats.isDirectory()) {
        dirs.push({ name: entry, path: fullPath, type: "dir" });
      } else {
        const ext = entry.slice(entry.lastIndexOf(".")).toLowerCase();
        if (VIDEO_EXTS.has(ext)) {
          files.push({ name: entry, path: fullPath, type: "file" });
        }
      }
    }
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "EACCES") {
      throw new HttpError(403, `Permission denied: ${target}`);
    }
    throw err;
  }

  const parent = dirname(target);
  return c.json({
    current: target,
    parent: parent !== target ? parent : null,
    dirs,
    files,
    video_count: files.length,
  });
});
